prime-router/src/main/kotlin/history/SubmissionHistory.kt

```diff
@@ -341,41 +341,61 @@ class DetailedSubmissionHistory(
+                        get(index).itemCountBeforeQualFilter =
+                            get(index).itemCountBeforeQualFilter?.plus(
+                                descendantDest.itemCountBeforeQualFilter ?: 0
+                            )
+                    } else {
+                        add(descendantDest)
+                    }
                 }
             }
         }
 
-        // if all the items were filtered out, no translate actions are generated (thus no destinations)
-        // to return the expected output, we create dummy destinations that are only used in enrichment
-        // these dummies are created using the receivers defined in the logs
-        if (destinations.isEmpty() && filterLogs.isNotEmpty()) {
-            val filteredReportRows = filterLogs.map { log -> log.detail.message }
-            val filteredReportItems = filterLogs.map { log ->
-                ReportStreamFilterResultForResponse(log.detail as ReportStreamFilterResult)
-            }
+        // local function to safe add elements to nullable immutable list
+        fun <T> List<T>?.add(element: T): List<T> {
+            val tempMutableList = this?.toMutableList() ?: arrayListOf()
+            tempMutableList.add(element)
+            return tempMutableList
+        }
+        // add filter logs to its respective destination otherwise add new destination
+        if (filterLogs.isNotEmpty()) {
             filterLogs.forEach { log ->
-                val filterResult = log.detail as ReportStreamFilterResult
-                val receiverNameSegments = filterResult.receiverName.split(Sender.fullNameSeparator)
-                destinations.add(
-                    Destination(
-                        receiverNameSegments[0],
-                        receiverNameSegments[1],
-                        filteredReportRows,
-                        filteredReportItems,
-                        null,
-                        0,
-                        filterResult.originalCount,
-                    )
-                )
+                with(destinations) {
+                    val filterResult = log.detail as ReportStreamFilterResult
+                    val filterReport = log.detail.message
+                    val receiverNameSegments = filterResult.receiverName.split(Sender.fullNameSeparator)
+                    val index = indexOfFirst {
+                        it.organizationId == receiverNameSegments[0] && it.service == receiverNameSegments[1]
+                    }
+                    if (index >= 0) {
+                        get(index).filteredReportRows =
+                            get(index).filteredReportRows.add(filterReport)
+                        get(index).filteredReportItems =
+                            get(index).filteredReportItems.add(ReportStreamFilterResultForResponse(filterResult))
+                        get(index).itemCountBeforeQualFilter = get(index).itemCountBeforeQualFilter?.plus(
```

Comment:
If I'm reading this correctly, if `itemCountBeforeQualFilter` is null, it'll stay null. Is that what we want, or should we set it to 1 in that case?

Reply:
This is increasing the existing `itemCountBeforeQualFilter` by whatever `itemCountBeforeQualFilter` is on the next destination. itemCountBeforeQualFilter can be null on the next destination and you can't add null so have to add 0 if null.

Reply:
Gotcha, let me start a new convo with the updated code, still have a question on this.